third_party/move/move-compiler-v2/src/cyclic_instantiation_checker.rs

```diff
@@ -0,0 +1,242 @@
+    fn visit_call(
+        &self,
+        nid: &NodeId,
+        op: &Operation,
+        insts: Vec<Type>,
+        callers_chain: &mut Vec<(Loc, QualifiedInstId<FunId>)>,
+    ) -> bool {
+        if let Operation::MoveFunction(mod_id, fun_id) = op {
+            let callee_uninst = mod_id.qualified_inst(*fun_id, self.get_inst(*nid));
+            let callee = callee_uninst.instantiate(&insts);
+            if *mod_id != self.mod_env.get_id() || self.def_not_recursive(callee.to_qualified_id())
+            {
+                // skips if callee from another module (since there is no cyclic module dependency),
+                // or if the callee is not recursive
+                true
+            } else {
+                for (_, ancester_caller) in callers_chain.iter() {
+                    if ancester_caller.to_qualified_id() == callee.to_qualified_id() {
+                        // we are checking for the root caller
+                        let (_, checking_for) = &callers_chain[0];
+                        if checking_for.to_qualified_id() != callee.to_qualified_id() {
+                            // check and report diagnostics when `callee` is checked
+                            // this happens when root caller `f` calls `g` which then calls `g` itself
+                            return true;
+                        } else {
+                            #[allow(clippy::collapsible_else_if)]
+                            if let Some(_ty_param) = callee
+                                .inst
+                                .iter()
+                                .filter_map(ty_properly_contains_ty_parameter)
+                                .next()
+                            {
+                                self.report_error(*nid, callee, callers_chain);
+                                return false;
+                            } else {
+                                return true;
+                            }
+                        }
+                    }
+                }
+                self.visit_callees(*nid, callee, insts, callers_chain)
+            }
+        } else {
+            true
+        }
+    }
+
+    /// Visits a call expression and checks for cyclic type instantiations.
+    /// Other parameters are the same as in `visit`.
+    /// Precondition: `caller` defined in `self.mod_env`
+    fn visit_callees(
+        &self,
+        caller_node: NodeId,
+        caller: QualifiedInstId<FunId>,
+        insts: Vec<Type>,
+        callers_chain: &mut Vec<(Loc, QualifiedInstId<FunId>)>,
+    ) -> bool {
+        let fun_env = self.mod_env.get_function(caller.id);
+        if let Some(caller_body) = fun_env.get_def() {
+            let caller_loc = self.mod_env.env.get_node_loc(caller_node);
+            callers_chain.push((caller_loc, caller));
+            let insts = Type::instantiate_vec(self.get_inst(caller_node), &insts);
+            let res = caller_body.visit_positions(&mut |pos, exp| {
```

Comment:
You are cloning `insts` for every subexpression, even when you do nothing (e.g., any non-`Call`).  Instead, you should pass `&insts` to the `visit` function and do a clone if you need it there.  Oh, but you don't even need an object, since in the end you use a reference anyway.  So use borrowed values instead.

Reply:
done